### PR TITLE
Add Tensor name to __str__ in Tensor Class

### LINKED ISSUE #453: Add Tensor name to __str__ and __repr__ in Tensor Class
🐞 **What bugs you?**

Currently, both `str` and `repr`,  in `einsteinpy.symbolic.tensor.py > Tensor`, don't show the Tensor name. As such, there is only one way to check the name, that is by explicitly accessing `tensor.name`.

🎯 **Goal**

Technically, `str` is supposed to be readable and `repr` is supposed to be unambiguous. Outputting the tensor `name` should improve both.

💡 **Possible solutions**

Add `name` to `str` and  `repr`. For example, the output of `str(tensor)` could be something like:
```
Tensor TensorName 
[[...]]
```

📋  **Steps to solve the problem**

 * Comment below about what you've started working on.
 * Add, commit, push your changes
 * Submit a pull request and add this in comments - `Addresses #<put issue number here>`
 * Ask for a review in comments section of pull request
 * Celebrate your contribution to this project 🎉

### PR DESCRIPTION
**Pull request by Starninja666**
Fixes #453 

@shreyasbapat  @ritzvik

### CODE REVIEW

**Comment by shreyasbapat** [comment]
Thanks for the Pull Request! I will review in a while.

File: src/einsteinpy/symbolic/tensor.py [hunk 1]
```diff
@@ -67,7 +67,7 @@ class Tensor:
     Base Class for Tensor manipulation
     """
 
-    def __init__(self, arr, config="ll"):
+    def __init__(self, arr, config="ll", name=None):
```
Comment (shreyasbapat) [inline src/einsteinpy/symbolic/tensor.py]: Hey @Starninja666 I guess you misunderstood the issue?

Comment (ritzvik) [reply src/einsteinpy/symbolic/tensor.py]: I know what the problem is. The `name` parameter is defined in `BaseRelativityTensor` while `__str__()` is there in `Tensor` :D

We would need to move `name` parameter to `Tensor` or `__str__()` to the derived class, either way it works. But personally, I like the first solution.

Comment (shreyasbapat) [reply src/einsteinpy/symbolic/tensor.py]: First solution anyday!

Comment (shreyasbapat) [reply src/einsteinpy/symbolic/tensor.py]: No. The motive is to put in the name in `__str__` and not just updating the  name variable!

Comment (shreyasbapat) [reply src/einsteinpy/symbolic/tensor.py]: Read this: https://www.journaldev.com/22460/python-str-repr-functions for more information about what is a string representation of a class.

Comment (ritzvik) [reply src/einsteinpy/symbolic/tensor.py]: I’ll have an extensive look tomorrow about what can be done

**Comment by ritzvik** [comment]
https://github.com/einsteinpy/einsteinpy/blob/b8351acb11eae83d55c8b861590fe8e16b82f768/src/einsteinpy/symbolic/tensor.py#L263-L267

Here, offload the responsibility of `name` parameter to `Tensor` class by `super(BaseRelativityTensor, self).__init__(arr=arr, config=config, name=name)` and remove the line `self.name=name`. 

Also, change the default `None` into `name="GenericTensor"` in `BaseRelativityTensor`. Update the docstrings accordingly

File: src/einsteinpy/symbolic/tensor.py [hunk 2]
```diff
@@ -77,7 +77,8 @@ def __init__(self, arr, config="ll"):
             Sympy Array, multi-dimensional list containing Sympy Expressions, or Sympy Expressions or int or float scalar
         config : str
             Configuration of contravariant and covariant indices in tensor. 'u' for upper and 'l' for lower indices. Defaults to 'll'.
-
+        name : string or None
```
Comment (ritzvik) [inline src/einsteinpy/symbolic/tensor.py]:
```suggestion
        name : str or None
```

File: src/einsteinpy/symbolic/tensor.py [hunk 3]
```diff
@@ -77,7 +77,8 @@ def __init__(self, arr, config="ll"):
...
+            Name of the tensor.
```
Comment (ritzvik) [inline src/einsteinpy/symbolic/tensor.py]:
```suggestion
            Name of the tensor.
            
```
Just a newline

File: src/einsteinpy/symbolic/tensor.py [hunk 4]
```diff
@@ -100,6 +101,7 @@ def __init__(self, arr, config="ll"):
             raise TypeError(
                 "config is either not of type 'str' or does contain characters other than 'l' or 'u'"
             )
+        self.name = name
```
Comment (ritzvik) [inline src/einsteinpy/symbolic/tensor.py]: Yes, that's pretty much what I wanted

File: src/einsteinpy/symbolic/tensor.py [hunk 5]
```diff
@@ -125,7 +127,10 @@ def __str__(self):
         Returns a String with a readable representation of the object of class Tensor
 
         """
-        representation = "Tensor" + "\n"
+        representation = "Tensor"
+        if self.name:
```
Comment (ritzvik) [inline src/einsteinpy/symbolic/tensor.py]:
```suggestion
        if self.name is not None:
```

This is a bit more clear :)

File: src/einsteinpy/symbolic/tensor.py [hunk 6]
```diff
@@ -125,7 +127,10 @@ def __str__(self):
...
+            representation += " " + self.name
```
Comment (ritzvik) [inline src/einsteinpy/symbolic/tensor.py]:
```suggestion
            representation = " ".join((representation, self.name))
```
A little more pythonic

File: src/einsteinpy/symbolic/tensor.py (hunk 1 above)
Comment (ritzvik) [reply src/einsteinpy/symbolic/tensor.py]: I'm resolving this.
I suggested the few changes required in a separate review. No need to clutter :)

**Comment by ritzvik** [comment]
For quality checks
https://docs.einsteinpy.org/en/latest/dev_guide.html#code-linting

**Comment by ritzvik** [comment]
> https://github.com/einsteinpy/einsteinpy/blob/b8351acb11eae83d55c8b861590fe8e16b82f768/src/einsteinpy/symbolic/tensor.py#L263-L267
> 
> Here, offload the responsibility of `name` parameter to `Tensor` class by `super(BaseRelativityTensor, self).__init__(arr=arr, config=config, name=name)` and remove the line `self.name=name`.
> 
> Also, change the default `None` into `name="GenericTensor"` in `BaseRelativityTensor`. Update the docstrings accordingly

No need to change default name. It's already done in another PR. Just rebase

**Comment by Starninja666** [comment]
All requested changes have been made, please review! :)

**Comment by ritzvik** [comment]
Code coverage is low because the tests are not hitting the `if` condition.

https://github.com/einsteinpy/einsteinpy/blob/0151c00e292abec7a172705196ec89e276e52804/src/einsteinpy/tests/test_symbolic/test_tensor.py#L113-L117

```
obj1 = Tensor(test_list, name="blah blah")
assert .... # whatever assert is already there
assert ("blah blah" in str(obj1))
```

This would fix it(probably)

**Comment by ritzvik** [comment]
After the changes, rebase :)

**Comment by ritzvik** [comment]
Thanks @Starninja666 
merging :)

**Pull request merged**